crates/rattler/src/repo_data/cache/mod.rs

```diff
@@ -0,0 +1,170 @@
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str(s)
+    }
+}
+
+/// Represents a value and when the value was last checked.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Expiring<T> {
+    pub value: T,
+
+    // #[serde(with = "chrono::serde::ts_seconds")]
+    pub last_checked: chrono::DateTime<chrono::Utc>,
+}
+
+impl<T> Expiring<T> {
+    pub fn value(&self, expiration: chrono::Duration) -> Option<&T> {
+        if chrono::Utc::now().signed_duration_since(self.last_checked) >= expiration {
+            None
+        } else {
+            Some(&self.value)
+        }
+    }
+}
+
+/// Deserializes a [`SystemTime`] by parsing an integer and converting that as a nanosecond based unix
+/// epoch timestamp to a [`SystemTime`].
+fn duration_from_nanos<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+    SystemTime::UNIX_EPOCH
+        .checked_add(std::time::Duration::from_nanos(Deserialize::deserialize(
+            deserializer,
+        )?))
+        .ok_or_else(|| D::Error::custom("the time cannot be represented internally"))
+}
+
+/// Serializes a [`SystemTime`] by converting it to a nanosecond based unix epoch timestamp.
+fn duration_to_nanos<S: Serializer>(time: &SystemTime, s: S) -> Result<S::Ok, S::Error> {
+    use serde::ser::Error;
+    time.duration_since(SystemTime::UNIX_EPOCH)
+        .map_err(|_| S::Error::custom("duration cannot be computed for file time"))?
+        .as_nanos()
+        .serialize(s)
+}
+
+fn deserialize_blake2_hash<'de, D>(
+    deserializer: D,
+) -> Result<Option<blake2::digest::Output<blake2::Blake2s256>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+    match Option::<&'de str>::deserialize(deserializer)? {
+        Some(str) => {
+            let mut hash = <blake2::digest::Output<blake2::Blake2s256>>::default();
+            hex::decode_to_slice(str, &mut hash).map_err(D::Error::custom)?;
+            Ok(Some(hash))
+        }
+        None => Ok(None),
+    }
+}
+
+fn serialize_blake2_hash<S: Serializer>(
+    time: &Option<blake2::digest::Output<blake2::Blake2s256>>,
+    s: S,
+) -> Result<S::Ok, S::Error> {
+    match time.as_ref() {
+        None => s.serialize_none(),
+        Some(hash) => format!("{:x}", hash).serialize(s),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::RepoDataState;
+    use std::str::FromStr;
+
+    #[test]
+    pub fn test_parse_repo_data_state() {
+        insta::assert_yaml_snapshot!(RepoDataState::from_str(
+            r#"{
+            "cache_control": "public, max-age=1200",
+            "etag": "\"bec332621e00fc4ad87ba185171bcf46\"",
+            "has_zst": {
+                "last_checked": "2023-02-13T14:08:50Z",
+                "value": true
+            },
+            "mod": "Mon, 13 Feb 2023 13:49:56 GMT",
+            "mtime_ns": 1676297333020928000,
+            "size": 156627374,
```

Comment:
this doesn't test the blake2 hash yet, right?

Reply:
No this file comes directly from something installed with Mamba. I wanted to ensure that I can load that cache :)

Reply:
OK, got it!